### PR TITLE
Upgrades the MC

### PR DESCRIPTION
**Pull request by AffectedArc07**
## What Does This PR Do
Updates the MC to TG's latest standards, minus the init stages part because we dont care about those.

The main addition here is that of the punisher, which will force SSs that overran their tick to be delayed, which should help with the game loop in highpop, especially when atmos is starting to lose it a bit.

## Why It's Good For The Game
Upgrades people, upgrades

## Testing
SSs still init, runlevels still work as intended, 0 runtimes.

I am still going to TM this on prod.

## Changelog
:cl: AffectedArc07
add: THE PUNISHER
experiment: Upgraded the MC
/:cl:

### CODE REVIEW

File: code/controllers/failsafe.dm [hunk 1]
```diff
@@ -92,6 +111,63 @@ GLOBAL_REAL(Failsafe, /datum/controller/failsafe)
...
+/proc/delete_all_SS_and_recreate_master()
+	// You can break EVERYTHING with this
+	if(usr && !check_rights(R_MAINTAINER))
+		return
+	del(Master)
+	for(var/global_var in global.vars)
+		if (istype(global.vars[global_var], /datum/controller/subsystem))
+			del(global.vars[global_var])
+	. = Recreate_MC()
+	if (. == 1) //We were able to create a new master
```
Comment (hal9000PR) [inline code/controllers/failsafe.dm:164]: spacing

File: code/controllers/failsafe.dm [hunk 2]
```diff
@@ -92,6 +111,63 @@ GLOBAL_REAL(Failsafe, /datum/controller/failsafe)
...
+	// You can break EVERYTHING with this
+	if(usr && !check_rights(R_MAINTAINER))
+		return
+	del(Master)
+	var/list/subsytem_types = subtypesof(/datum/controller/subsystem)
+	sortTim(subsytem_types, /proc/cmp_subsystem_init)
+	for(var/I in subsytem_types)
+		new I
+	. = Recreate_MC()
+	if (. == 1) //We were able to create a new master
```
Comment (hal9000PR) [inline code/controllers/failsafe.dm:147]: spacing

File: code/controllers/failsafe.dm [hunk 3]
```diff
@@ -92,6 +111,63 @@ GLOBAL_REAL(Failsafe, /datum/controller/failsafe)
...
+		message_admins("<span class='boldannounce'>Failed to create new MC!</span>")
+
+///Delete all existing SS to basically start over
+/proc/delete_all_SS_and_recreate_master()
+	// You can break EVERYTHING with this
+	if(usr && !check_rights(R_MAINTAINER))
+		return
+	del(Master)
+	for(var/global_var in global.vars)
+		if (istype(global.vars[global_var], /datum/controller/subsystem))
```
Comment (hal9000PR) [inline code/controllers/failsafe.dm:161]: spacing

File: code/controllers/subsystem.dm [hunk 4]
```diff
@@ -215,24 +311,24 @@
...
 		if(SS_IDLE) // Leave it default if the SS is idle
 			. = "<font>"
-//could be used to postpone a costly subsystem for (default one) var/cycles, cycles
-//for instance, during cpu intensive operations like explosions
+
+/// Causes the next "cycle" fires to be missed. Effect is accumulative but can reset by calling update_nextfire(reset_time = TRUE)
 /datum/controller/subsystem/proc/postpone(cycles = 1)
-	if(next_fire - world.time < wait)
-		next_fire += (wait*cycles)
+	if (can_fire && cycles >= 1)
```
Comment (hal9000PR) [inline code/controllers/subsystem.dm:317]: spacing

File: code/controllers/subsystem.dm [hunk 5]
```diff
@@ -215,24 +311,24 @@
...
 
 //usually called via datum/controller/subsystem/New() when replacing a subsystem (i.e. due to a recurring crash)
 //should attempt to salvage what it can from the old instance of subsystem
 /datum/controller/subsystem/Recover()
 
 /datum/controller/subsystem/vv_edit_var(var_name, var_value)
-	switch(var_name)
-		if("can_fire")
+	switch (var_name)
+		if (NAMEOF(src, can_fire))
```
Comment (hal9000PR) [inline code/controllers/subsystem.dm:326]: spacing

File: code/controllers/subsystem.dm [hunk 6]
```diff
@@ -215,24 +311,24 @@
...
 			//this is so the subsystem doesn't rapid fire to make up missed ticks causing more lag
-			if(var_value)
-				next_fire = world.time + wait
-		if("queued_priority") //editing this breaks things.
-			return 0
+			if (var_value)
+				update_nextfire(reset_time = TRUE)
+		if (NAMEOF(src, queued_priority)) //editing this breaks things.
```
Comment (hal9000PR) [inline code/controllers/subsystem.dm:330]: ditto

**Review by hal9000PR — APPROVED** [review]

**Comment by AffectedArc07** [comment]
TM'd with 0 issues

**Pull request merged**